### DISCUSSION TITLE
Remote upgrade linux doesn't work

### DISCUSSION THREAD

**Discussion by mattuschek** [topgrade-rs/topgrade#716] [Q&A] [answered]
Hello,

I am topgrade Newbie.

I have a Proxmox host with LXC container.
I want remote unattended upgrade the LXC container (debian linux).
But it doesn't work.
Where is the problem?

Below is the description of what I did.

topgrade was installed on Proxmox host.
```
apt install tmux -y
wget https://github.com/topgrade-rs/topgrade/releases/download/v14.0.1/topgrade-v14.0.1-x86_64-unknown-linux-gnu.tar.gz
tar -vxzf topgrade-v14.0.1-x86_64-unknown-linux-gnu.tar.gz
mv topgrade /usr/local/bin
chmod +x /usr/local/bin/topgrade
# topgrade copied to LXC container
scp /usr/local/bin/topgrade 192.168.1.85:/usr/local/bin/topgrade
```

Config-File on Proxmox host.
```
root@pve83:~ # cat ~/.config/topgrade.toml
[misc]

# List of remote machines with Topgrade installed on them
remote_topgrades = ["192.168.1.85"]

# Do not set the terminal title (default: true)
set_title = false

# Don't ask for confirmations (no default value)
assume_yes = true

# Do not ask to retry failed steps (default: false)
no_retry = true

# Run inside tmux (default: false)
# run_in_tmux = true
```

SSH connection to LXC container works.
```
root@pve83:~ # ssh root@192.168.1.85
Linux debian85 6.5.11-8-pve #1 SMP PREEMPT_DYNAMIC PMX 6.5.11-8 (2024-01-30T12:27Z) x86_64

The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
permitted by applicable law.
Last login: Sat Feb 24 14:11:45 2024 from 192.168.23.27
root@debian85:~ #
```

topgrade is started on Proxmox host WITHOUT temux.
No matter, whether with `topgrade` or `topgrade -y`.
Result is the same.
```
root@pve83:~ # topgrade -y

-- 18:09:02 - Self update ------------------------------------------------------
Checking target-arch... x86_64-unknown-linux-gnu
Checking current version... v14.0.1
Checking latest released version... v14.0.1
Topgrade is up-to-date

-- 18:09:02 - Remote (192.168.1.85) -------------------------------------------
Connecting to 192.168.1.85...

-- (192.168.1.85) 18:09:02 - Self update --------------------------------------
Checking target-arch... x86_64-unknown-linux-gnu
Checking current version... v14.0.1
Checking latest released version... v14.0.1
Topgrade is up-to-date

-- (192.168.1.85) 18:09:02 - System update ------------------------------------
OK:1 http://security.debian.org bookworm-security InRelease
OK:2 http://deb.debian.org/debian bookworm InRelease
OK:3 http://deb.debian.org/debian bookworm-updates InRelease
OK:4 https://apt.syncthing.net syncthing InRelease
Paketlisten werden gelesen… Fertig
Paketlisten werden gelesen… Fertig
Abhängigkeitsbaum wird aufgebaut… Fertig
Statusinformationen werden eingelesen… Fertig
Paketaktualisierung (Upgrade) wird berechnet… Fertig
Die folgenden Pakete werden aktualisiert (Upgrade):
  base-files bind9-dnsutils bind9-host bind9-libs libc-bin libc-l10n libc6 libcryptsetup12 libgnutls30 libmariadb3 libnss-systemd libpam-systemd libpq5
  libsystemd-shared libsystemd0 libudev1 locales mariadb-common syncthing systemd systemd-sysv systemd-timesyncd tar tzdata udev usrmerge
26 aktualisiert, 0 neu installiert, 0 zu entfernen und 0 nicht aktualisiert.
Es müssen 31,1 MB an Archiven heruntergeladen werden.
Nach dieser Operation werden 193 kB Plattenplatz zusätzlich benutzt.
Möchten Sie fortfahren? [J/n]
```

**Comment by SteveLauC** [comment]
> But it doesn't work.


From the log, it seems that Topgrade is working in the remote container?

> -- 18:09:02 - Remote (192.168.1.85) -------------------------------------------

**Reply by mattuschek** [reply]
Where is the Topgrade log?
And sorry, I mean an remote unattended upgrade.

**Reply by SteveLauC** [reply]
> Where is the Topgrade log?

Sorry, I mean the output.

>  I mean an remote unattended upgrade.

Do you mean that you want to the assume_yes config work with the remote updates?

> Möchten Sie fortfahren? [J/n]

so that you won't need to input `J` yourself?

**Reply by SteveLauC** [reply]
This is indeed something Topgrade does not handle, see #558.

Currently, you can only do it by putting a config file in your container, where the `assume_yes` option is enabled.

**Reply by mattuschek** [reply]
> > Where is the Topgrade log?
> 
> Sorry, I mean the output.

OK.
But still the question, is there a topgrade log somewhere?

> > I mean an remote unattended upgrade.
> 
> Do you mean that you want to the assume_yes config work with the remote updates?

Yes, that I mean.

> > Möchten Sie fortfahren? [J/n]
> 
> so that you won't need to input `J` yourself?

Yes.

> This is indeed something Topgrade does not handle, see #558.
> 
> Currently, you can only do it by putting a config file in your container, where the `assume_yes` option is enabled.

Thanks for info.
OK,  I unterstand. By default, Topgrade cannot perform a remote unattended upgrade.
Is it possible that in the future Topgrade will copy the config file itself to the remote computer?
That would be a good improvement.

**Reply by SteveLauC** [reply]
> But still the question, is there a topgrade log somewhere?

Topgrade directly prints logs to stdout, you should be able to see them if the log level has been adjusted appropriately, e.g., setting log level to `trace`

```sh
$ topgrade --log-filte trace
```

> By default, Topgrade cannot perform a remote unattended upgrade.
Is it possible that in the future Topgrade will copy the config file itself to the remote computer?
That would be a good improvement.

I don't have a plan for this as it can cause problems, take remote execution as an example:

```toml
# List of remote machines with Topgrade installed on them
remote_topgrades = ["192.168.1.85"]
```

If topgrade copies this config file to your container, then the topgrade within your container will also try to do the remote execution, which is obviously not what we want.

But I think we can implement the features requested in #558, passing cli argument to remote executions.

**Reply by mattuschek** [reply]
> > By default, Topgrade cannot perform a remote unattended upgrade.
> 
> I don't have a plan for this as it can cause problems, take remote execution as an example:
> 
> ```toml
> # List of remote machines with Topgrade installed on them
> remote_topgrades = ["192.168.1.85"]
> ```
> 
> If topgrade copies this config file to your container, then the topgrade within your container will also try to do the remote execution, which is obviously not what we want.

Yes, you're right. That's a problem.

> But I think we can implement the features requested in #558, passing cli argument to remote executions.

Good idea.

Either a cli argument.
Example:
`--remote_yes`

Or a parameter in the config file `topgrade.toml`.
Example:
```
# Say yes to package manager's prompt of remote machines
remote_yes = true

```

**Reply by SteveLauC** [reply]
If we implement the feature stated in [this comment](https://github.com/topgrade-rs/topgrade/issues/558#issuecomment-1742016576), then we don't need a dedicated `--remote-yes` option, simply executing `topgrade -y` on host would work:)

**Reply by mattuschek** [reply]
By when could this be implemented?

**Reply by SteveLauC** [reply]
I will give it a rethink and probably do it before the next release.

**Reply by SteveLauC** [reply]
I just realized that the `-y` option requires you to input step names, `topgrade -y` (no step name given) won't enable the assume_yes option for any steps, you still have to do it in the container's config file.

**Reply by mattuschek** [reply, marked as answer]
Now the remote systems have their own configuration file `topgrade.toml`.
I hope that the remote systems will no longer prompt for "J" (or "Y") in the future.

```
root@debian85:~ # cat ~/.config/topgrade.toml
[misc]

# Do not set the terminal title (default: true)
set_title = false

# Don't ask for confirmations (no default value)
assume_yes = true

# Do not ask to retry failed steps (default: false)
no_retry = true

# Run inside tmux (default: false)
#run_in_tmux = true
```
